src/dbnode/storage/flush.go

```diff
@@ -102,28 +105,58 @@ func (m *flushManager) Flush(
 
 	defer m.setState(flushManagerIdle)
 
-	// create flush-er
-	flushPersist, err := m.pm.StartFlushPersist()
+	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
 		return err
 	}
 
-	namespaces, err := m.database.GetOwnedNamespaces()
+	// Perform three separate loops through all the namespaces so that we can
+	// emit better gauges, i.e. all the flushing for all the namespaces happens
+	// at once, then all the cold flushes, then all the snapshotting. This is
+	// also slightly better semantically because flushing should take priority
+	// over cold flushes and snapshotting.
+	//
+	// In addition, we need to make sure that for any given shard/blockStart
+	// combination, we attempt a flush before a snapshot as the snapshotting
+	// process will attempt to snapshot any unflushed blocks which would be
+	// wasteful if the block is already flushable.
+	multiErr := xerrors.NewMultiError()
+	if err = m.dataWarmFlush(namespaces, tickStart, dbBootstrapStateAtTickStart); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	rotatedCommitlogID, err := m.commitlog.RotateLogs()
+	if err == nil {
+		if err = m.dataColdFlush(namespaces); err != nil {
```

Comment:
Yeah cold flush is dependent on commitlog rotation because snapshots only capture warm writes. Both a cold flush and snapshot is required to capture all data in the rotated commitlog.